include/boost/gil/image_processing/morphology.hpp

```diff
@@ -0,0 +1,285 @@
+          row_boundary = view_row + (kernel.center_y() - flip_ker_row);
+          col_boundary = view_col + (kernel.center_x() - flip_ker_col);
+
+          // ignore input samples which are out of bound
+          if (row_boundary >= 0 && row_boundary < src_view.height() &&
+              col_boundary >= 0 && col_boundary < src_view.width()) {
+            // We ensure that we consider only those pixels which are overlapped
+            // on a non-zero kernel_element as
+            if(kernel.at(flip_ker_row,flip_ker_col) == 0){
+              continue;
+            }
+            if (identifier == morphological_operation::dilation) {
+              if (src_view(col_boundary, row_boundary) > target_element)
+                target_element = src_view(col_boundary, row_boundary);
+            } else if (identifier == morphological_operation::erosion) {
+              if (src_view(col_boundary, row_boundary) < target_element)
+                target_element = src_view(col_boundary, row_boundary);
+            }
+          }
+        }
+      }
+      dst_view(view_col, view_row) = target_element;
+    }
+  }
+}
+
+/// \brief Checks feasibility of the desired operation and passes parameter
+/// values to the function morph_impl alongwith individual channel views of the
+/// input image. \param src_view - Source/Input image view. \param dst_view -
+/// View which stores the final result of operations performed by this function.
+/// \param kernel - Kernel matrix/structuring element containing 0's and 1's
+/// which will be used for applying the required morphological operation. \param
+/// identifier - Indicates the type of morphological operation to be applied.
+/// \tparam SrcView type of source image.
+/// \tparam DstView type of output image.
+/// \tparam Kernel type of structuring element.
+template <typename SrcView, typename DstView, typename Kernel>
+void morph(SrcView const &src_view, DstView const &dst_view, Kernel const &ker_mat,
+           morphological_operation identifier) {
+  BOOST_ASSERT(ker_mat.size() != 0 &&
+               src_view.dimensions() == dst_view.dimensions());
+  gil_function_requires<ImageViewConcept<SrcView>>();
+  gil_function_requires<MutableImageViewConcept<DstView>>();
+
+  gil_function_requires<ColorSpacesCompatibleConcept<
+                              typename color_space_type<SrcView>::type,
+                              typename color_space_type<DstView>::type>>();
+
+  using d_channel_t = typename channel_type<SrcView>::type;
+  using channel_t = typename channel_convert_to_unsigned<d_channel_t>::type;
+  using gray_pixel_t = pixel<channel_t, gray_layout_t>;
+  using src_gray_image = image<gray_pixel_t, false>;
+  src_gray_image intermediate_img(src_view.dimensions());
+
+  for (std::size_t i = 0; i < src_view.num_channels(); i++) {
+    morph_impl(nth_channel_view(src_view, i),
+               nth_channel_view(view(intermediate_img), i), ker_mat,
```

Comment:
Intermediate image is only one channel, but `i` might take values higher than 0. Please try to test it with multi-channel image.

Reply:
Yes you are correct, it was creating problems with multi-channelled images, I changed the declaration of intermediate_img as 
```
gil::image<typename DstView::value_type> intermediate_img(src_view.dimensions());
```
Now,it is working fine even with multi-channelled images.Thanks for pointing that out.